File: 11-container-with-most-water/11-container-with-most-water.cpp
class Solution {
public:
    int min(int a, int b){
        if (a<b) return a;
        return b;
    }
    int maxArea(vector<int>& height) {
        int len = height.size();
        long long res = -1, area;
        int p1 = 0, p2 = len - 1;
        
        while(p1 < len && p2 >= 0){
            area = min(height[p1], height[p2]) * (p2 - p1);
            if(area > res) res = area;
            
            if (height[p1] <= height[p2]){
                p1++;
            }
            else p2--;
        }
        
        return res;
    }
};

File: 11-container-with-most-water/11-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        l=0
        r=len(height)-1
        res=0
        
        while l<r:
            water=(r-l)*min(height[l], height[r])
            
            if water>res: res=water
            
            if height[l] > height[r]:
                r-=1
            elif height[l] < height[r]:
                l+=1
            else:
                l+=1
                r-=1
        
        return res